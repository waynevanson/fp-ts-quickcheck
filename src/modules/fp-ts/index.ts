export * as state from "./state"
export * as stateT from "./stateT"
export * as task from "./task"
export * as io from "./io"
