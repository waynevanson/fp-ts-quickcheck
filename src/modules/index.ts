export * as iterable from "./iterable"
export * as generator from "./generator"
export * as lcg from "./lcg"
