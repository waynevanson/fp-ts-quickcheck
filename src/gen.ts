import * as gen from "@no-day/fp-ts-generators"
import * as lcg from "@no-day/fp-ts-lcg"
import { nonEmptyArray, state as S } from "fp-ts"
import { chainFirst as _chainFirst } from "fp-ts/lib/Chain"
import { Lazy, pipe } from "fp-ts/lib/function"
import * as lens from "monocle-ts/Lens"

export * from "@no-day/fp-ts-generators"

export const chainFirst = _chainFirst(gen.Monad)

export function variant(seed: number): gen.Gen<void> {
  return S.modify(
    pipe(
      lens.id<gen.GenState>(),
      lens.prop("newSeed"),
      lens.modify(() => lcg.mkSeed(seed)),
    ),
  )
}

export const nextSeed: gen.Gen<void> = S.modify(({ newSeed, size }) => ({
  size,
  newSeed: lcg.lcgNext(newSeed),
}))

export function union<T extends readonly [unknown, ...(readonly unknown[])]>(
  ...gens: { readonly [P in keyof T]: gen.Gen<T[P]> }
): gen.Gen<T[number]> {
  return gen.oneOf(
    gens as unknown as nonEmptyArray.NonEmptyArray<gen.Gen<T[number]>>,
  )
}

export const of: <A>(a: A) => gen.Gen<A> = gen.of

export const map: <A, B>(f: (a: A) => B) => (fa: gen.Gen<A>) => gen.Gen<B> =
  gen.map

export const ap: <A>(
  fa: gen.Gen<A>,
) => <B>(fab: gen.Gen<(a: A) => B>) => gen.Gen<B> = gen.ap

export function lazy<A>(lazy: Lazy<gen.Gen<A>>): gen.Gen<A> {
  return (s) => lazy()(s)
}
