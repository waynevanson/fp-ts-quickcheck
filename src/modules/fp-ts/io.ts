import { io as IO } from "fp-ts"
import { MonadRecIO1 } from "../monad-rec-io"

export * from "fp-ts/IO"

export const MonadRecIO: MonadRecIO1<IO.URI> = {
  ...IO.ChainRec,
  ...IO.FromIO,
  ...IO.Pointed,
}
