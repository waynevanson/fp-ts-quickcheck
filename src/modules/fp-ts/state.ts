import * as E from "fp-ts/Either"
import { ChainRec2 } from "fp-ts/lib/ChainRec"
import { Chain, URI, State } from "fp-ts/State"

export * from "fp-ts/State"

export const ChainRec: ChainRec2<URI> = {
  ...Chain,
  chainRec: (a, f) => (s) => {
    // eslint-disable-next-line functional/no-let
    let [_ea, _state] = f(a)(s)

    // eslint-disable-next-line functional/no-loop-statement
    while (E.isLeft(_ea)) {
      const result = f(_ea.left)(_state)
      // eslint-disable-next-line functional/no-expression-statement
      _ea = result[0]
      // eslint-disable-next-line functional/no-expression-statement
      _state = result[1]
    }

    return [_ea.right, _state]
  },
}

export const chainRec =
  <A, B, S>(f: (a: A) => State<S, E.Either<A, B>>) =>
  (a: A) =>
    ChainRec["chainRec"](a, f)
