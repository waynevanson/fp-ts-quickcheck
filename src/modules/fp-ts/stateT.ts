export * from "fp-ts/StateT"

import { either as E, stateT as ST, chain as CH } from "fp-ts"
import { URIS } from "fp-ts/HKT"
import { ChainRec, ChainRec1 } from "fp-ts/lib/ChainRec"
import { Endomorphism } from "fp-ts/lib/Endomorphism"
import { constVoid, identity, pipe } from "fp-ts/lib/function"
import { Pointed, Pointed1 } from "fp-ts/lib/Pointed"

export function gets<F extends URIS>(
  P: Pointed1<F>,
): <S, A>(f: (s: S) => A) => ST.StateT1<F, S, A>
export function gets<F>(
  P: Pointed<F>,
): <S, A>(f: (s: S) => A) => ST.StateT<F, S, A>
export function gets<F>(
  P: Pointed<F>,
): <S, A>(f: (s: S) => A) => ST.StateT<F, S, A> {
  return (f) => (s) => P.of([f(s), s])
}

export const get =
  <F extends URIS>(P: Pointed1<F>) =>
  <S>(): ST.StateT1<F, S, S> =>
    gets(P)(identity)

export function chainFirst<F extends URIS>(
  F: CH.Chain1<F>,
): <A, S, B>(
  f: (a: A) => ST.StateT1<F, S, B>,
) => (fa: ST.StateT1<F, S, A>) => ST.StateT1<F, S, A>
export function chainFirst<F>(
  F: CH.Chain<F>,
): <A, S, B>(
  f: (a: A) => ST.StateT<F, S, B>,
) => (fa: ST.StateT<F, S, A>) => ST.StateT<F, S, A>
export function chainFirst<F>(
  F: CH.Chain<F>,
): <A, S, B>(
  f: (a: A) => ST.StateT<F, S, B>,
) => (fa: ST.StateT<F, S, A>) => ST.StateT<F, S, A> {
  return (f) => (fa) =>
    pipe(
      fa,
      ST.chain(F)((a) =>
        pipe(
          f(a),
          ST.map(F)(() => a),
        ),
      ),
    )
}

export function chainRec<F extends URIS>(
  M: ChainRec1<F>,
): <S, A, B>(
  a: A,
  f: (a: A) => ST.StateT1<F, S, E.Either<A, B>>,
) => ST.StateT1<F, S, B>
export function chainRec<F>(
  M: ChainRec<F>,
): <S, A, B>(
  a: A,
  f: (a: A) => ST.StateT<F, S, E.Either<A, B>>,
) => ST.StateT<F, S, B>
export function chainRec<F>(
  M: ChainRec<F>,
): <S, A, B>(
  a: A,
  f: (a: A) => ST.StateT<F, S, E.Either<A, B>>,
) => ST.StateT<F, S, B> {
  return (a, f) => (s) =>
    M.chainRec([a, s] as const, ([a, s]) =>
      M.map(f(a)(s), ([ea, s]) =>
        pipe(
          ea,
          E.bimap(
            (e) => [e, s] as const,
            (a) => [a, s],
          ),
        ),
      ),
    )
}

export function modify<F extends URIS>(
  F: Pointed1<F>,
): <S>(f: Endomorphism<S>) => ST.StateT1<F, S, void>
export function modify<F>(
  F: Pointed<F>,
): <S>(f: Endomorphism<S>) => ST.StateT<F, S, void>
export function modify<F>(
  F: Pointed<F>,
): <S>(f: Endomorphism<S>) => ST.StateT<F, S, void> {
  return (f) => (s) => F.of([constVoid(), f(s)])
}

// export function apS<F extends URIS>(F: Apply1<F>) {
//   return <N extends string, A, E, B>(
//       name: Exclude<N, keyof A>,
//       fb: ST.StateT1<F, E, B>,
//     ) =>
//     (
//       fa: ST.StateT1<F, E, A>,
//     ): ST.StateT1<
//       F,
//       E,
//       { readonly [K in N | keyof A]: K extends keyof A ? A[K] : B }
//     > => {
//       return
//     }
// }
