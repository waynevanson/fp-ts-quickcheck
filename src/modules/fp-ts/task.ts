import { task as T } from "fp-ts"
import { ChainRec1 } from "fp-ts/lib/ChainRec"
import { tailRecM } from "../../utils"

export * from "fp-ts/Task"

export const ChainRec: ChainRec1<T.URI> = {
  ...T.Chain,
  chainRec: (fa, f) => tailRecM(T.Monad)(f)(T.of(fa)),
}
