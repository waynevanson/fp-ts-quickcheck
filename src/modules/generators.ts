import * as gen from "@no-day/fp-ts-generators"
import * as lcg from "@no-day/fp-ts-lcg"
import { state as S, monoid as M, number as Number } from "fp-ts"
import { chainFirst as _chainFirst } from "fp-ts/lib/Chain"
import { pipe } from "fp-ts/lib/function"
import * as lens from "monocle-ts/Lens"

export * from "@no-day/fp-ts-generators"

export const chainFirst = _chainFirst(gen.Monad)

export function variant(seed: number): gen.Gen<void> {
  return S.modify(
    pipe(
      lens.id<gen.GenState>(),
      lens.prop("newSeed"),
      lens.modify(() => lcg.mkSeed(seed)),
    ),
  )
}

export const nextSeed: gen.Gen<void> = S.modify(({ newSeed, size }) => ({
  size,
  newSeed: lcg.lcgNext(newSeed),
}))
