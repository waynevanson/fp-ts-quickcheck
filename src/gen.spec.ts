import { pipe } from "fp-ts/lib/function"
import * as gen from "./gen"
import { state } from "./modules/fp-ts"

describe("gen", () => {
  describe("variant", () => {
    it("should replace the current seed with the input from variant", () => {
      const newSeed = 24
      const result = pipe(
        gen.variant(newSeed),
        state.execute({ newSeed: gen.mkSeed(0), size: 0 }),
      )
      expect(result.newSeed).toEqual(gen.mkSeed(newSeed))
    })
  })

  describe("nextSeed", () => {
    it("should use a different seed from the one provided", () => {
      const newSeed = gen.mkSeed(24)
      const result = pipe(gen.nextSeed, state.execute({ newSeed, size: 0 }))
      expect(result.newSeed).not.toEqual(newSeed)
    })
  })

  describe("lazy", () => {
    it("should allow usage of generators before their declaration", () => {
      const genState = { newSeed: gen.mkSeed(43), size: 0 }
      const lazyGen = gen.lazy(() => nextGen)
      const value = 24
      const nextGen = gen.of(value)
      const result = pipe(lazyGen, state.evaluate(genState))
      expect(result).toBe(value)
    })
  })

  describe("nullable", () => {
    it("should generate null and the current generator", () => {
      const value = "value"
      const result = pipe(
        gen.of(value),
        gen.nullable,
        gen.generateSample({ seed: gen.mkSeed(23), count: 10 }),
      )
      expect(result).toMatchSnapshot()
    })
  })

  describe("filter", () => {
    it("should regenarate the values until a valid value has been found", () => {
      const predicate = (a: string) => a.length > 3
      const result = pipe(
        gen.string({ from: "a", to: "z" }),
        gen.filter(predicate),
        gen.generateSample({ seed: gen.mkSeed(24), count: 4 }),
      )

      expect(result.every(predicate)).toBeTruthy()
    })
  })
})
