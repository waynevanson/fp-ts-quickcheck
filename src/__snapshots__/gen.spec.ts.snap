// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gen nullable should generate null and the current generator 1`] = `
Array [
  null,
  null,
  null,
  null,
  null,
  null,
  null,
  "value",
  "value",
  null,
]
`;
